### PR TITLE
materializations: integration test improvements

### PR DESCRIPTION
**Pull request by williamhbaker**
**Description:**

PR is taken over by Mahdi, text below also by Mahdi:

- Rework integration tests completely so that instead of running mostly in bash scripts, they are now part of the Go test suite
- One important change here is that resources set up by integration tests are now unique to each run, so that multiple parallel runs of integration tests will not collide. This should improve our CI reliability a lot.
- The tests are a lot more comprehensive now, including many more edge cases previously not covered by integration tests. This has also led to some bugs being discovered in some connectors, hence the connector code changes.

**Workflow steps:**

(How does one use this feature, and how has it changed)

**Documentation links affected:**

(list any [documentation links](https://docs.google.com/document/d/1SRC9VS9zyCzWl3n4HXHbc4wPB1eLxJHkA2rtu9ZNokM/edit?usp=sharing) that you created, or existing ones that you've identified as needing updates, along with a brief description)

**Notes for reviewers:**

(anything that might help someone review this PR)

### CODE REVIEW

File: materialize-boilerplate/materializer.go [hunk 1]
```diff
@@ -300,6 +302,20 @@ type Materializer[
...
+	// in a checkpoints table.
+	ListTestTasks(context.Context) ([]string, error)
+
+	// CleanupTestTask performs any task-level cleanup actions that should be
+	// run after completing an integration test.
+	CleanupTestTask(context.Context, string) error
+
+	// SnapshotTestResource retrieves all rows of data contained in the
+	// materialized resource.
+	SnapshotTestResource(context.Context, []string) (columnNames []string, rows [][]any, _ error)
```
Comment (jacobmarble) [inline materialize-boilerplate/materializer.go]: Please name the `[]string` parameter so readers know what it represents. Similar request for the other methods in this interface.

File: materialize-azure-blob-parquet/Dockerfile [hunk 2]
```diff
@@ -16,7 +16,7 @@ COPY materialize-azure-blob-parquet    ./materialize-azure-blob-parquet
 
 RUN go test -v ./filesink/...
 RUN go test -v ./materialize-boilerplate/...
-RUN go test -v ./materialize-azure-blob-parquet/...
+RUN go test -short -v ./materialize-azure-blob-parquet/...
```
Comment (jacobmarble) [inline materialize-azure-blob-parquet/Dockerfile]: I notice that the `nodb` tag and `TEST_DATABASE` env var are entirely removed in this PR, replaced with short tests. That's a nice improvement! Perhaps the capture team would like make the same change to the `TEST_DATABASE` env var in the source connectors (in a subsequent PR).

**Review by jacobmarble — COMMENTED** [review]
It would be nice to (eventually) enforce the conventional patterns, like common sequences in Dockerfiles, through a lint tool of some kind.

File: tests/materialize/run.sh [hunk 3]
```diff
@@ -5,152 +5,30 @@ set -o pipefail
...
-    git diff --exit-code ${SNAPSHOT} || bail "Test Failed because snapshot is different"
+  # Test common materialization packages. This is repeated for each
+  # materialization during CI runs which is quite redundant, but these tests are
+  # fast so it doesn't really matter. This should be optimized later with other CI
+  # improvements.
+
+  cd $ROOT_DIR/go && go test -v ./...
+  cd $ROOT_DIR/filesink && go test -v ./...
+  cd $ROOT_DIR/materialize-boilerplate && go test -v ./...
+  go test -v ./... -timeout 20m
```
Comment (jacobmarble) [inline tests/materialize/run.sh]: As-is, the script runs `go test` in `$ROOT_DIR/materialize-boilerplate` twice. Is this what is meant?

```suggestion
  cd $ROOT_DIR/$CONNECTOR && go test -v ./... -timeout 20m
```

File: materialize-boilerplate/testutil/test_support.go [hunk 4]
```diff
@@ -107,6 +153,32 @@ func RunMaterializationTest[EC boilerplate.EndpointConfiger, FC boilerplate.Fiel
...
+// concurrently (up to maxParallelTasks at a time). Use this for connectors
+// where parallel task execution is safe and beneficial.
+func RunMaterializationTestParallel[EC boilerplate.EndpointConfiger, FC boilerplate.FieldConfiger, RC boilerplate.Resourcer[RC, EC], MT boilerplate.MappedTyper](
+	t *testing.T,
+	newMaterializer boilerplate.NewMaterializerFn[EC, FC, RC, MT],
+	sourcePath string,
+	makeResourceFn func(finalResourcePathPart string, deltaUpdates bool) RC,
+	actionDescSanitizers []func(string) string,
+) {
+	ctx := context.Background()
```
Comment (jacobmarble) [inline materialize-boilerplate/testutil/test_support.go:173]: nit: `t.Context()`, same comment elsewhere.

**Review by jacobmarble — APPROVED** [review]
It seems like you had a reason to switch from S3-compatible services RustFS and LocalStack to Minio.

**Comment by mdibaiee** [comment]
@jacobmarble LocalStack has become paid for certain features, and it was complaining on the latest version. So I switched to MinIO which is free

File: materialize-boilerplate/testutil/test_support.go (hunk 4 above)
Comment (mdibaiee) [reply materialize-boilerplate/testutil/test_support.go:173]: This is actually `context.Background` for a reason, since materializer / cleanups / snapshot and a few other steps outlive the test itself, so if we use `t.Context()`, context gets cancelled and we can't cleanup / snapshot, etc. -- I tested it

File: materialize-s3-parquet/docker-compose.yaml [hunk 5]
```diff
@@ -1,22 +1,24 @@
 services:
   storage:
-    image: rustfs/rustfs:1.0.0-alpha.90
-    pull_policy: missing
+    image: minio/minio:latest
```
Comment (danielnelson) [inline materialize-s3-parquet/docker-compose.yaml:3]: I moved this to rustfs, from localstack, the other day in https://github.com/estuary/connectors/pull/4066.

Unfortunately minio is unmaintained these days, so it's not a good replacment for localstack anymore.

Comment (mdibaiee) [reply materialize-s3-parquet/docker-compose.yaml:3]: RustFS was also complaining about a specific feature not being supported IIRC

Comment (danielnelson) [reply materialize-s3-parquet/docker-compose.yaml:3]: Hmm, I wonder what might have changed since #4066 was merged.  I say we don't worry about it too much and get this merged so its not a rebase headache, then we can switch back to rustfs or similar later.

**Review by danielnelson — APPROVED** [review]

File: materialize-s3-parquet/docker-compose.yaml (hunk 5 above)
Comment (mdibaiee) [reply materialize-s3-parquet/docker-compose.yaml:3]: Okay, I can try to re-visit after merging this, maybe I did something wrong while testing with RustFS
